File: 02_Object_Oriented_Programming/02_Class_Relationships/06_Realization/python/credit_card_payment.py
"""
Realization — CreditCardPayment implements Payment.
"""

from payment import Payment


class CreditCardPayment(Payment):
    """Pays using a credit card."""

    def __init__(self, card_number: str) -> None:
        self._card_number = card_number

    @property
    def payment_method(self) -> str:
        return "Credit Card"

    def pay(self, amount: float) -> None:
        masked = f"****-{self._card_number[-4:]}"
        print(f"Paid ${amount} using Credit Card ({masked}).")

File: 02_Object_Oriented_Programming/02_Class_Relationships/06_Realization/java/CreditCardPayment.java
/**
 * Realization — CreditCardPayment implements Payment.
 */
public class CreditCardPayment implements Payment {

    private String cardNumber;

    public CreditCardPayment(String cardNumber) {
        this.cardNumber = cardNumber;
    }

    @Override
    public void pay(double amount) {
        String masked = "****-" + cardNumber.substring(cardNumber.length() - 4);
        System.out.println("Paid $" + amount + " using Credit Card (" + masked + ").");
    }

    @Override
    public String getPaymentMethod() {
        return "Credit Card";
    }
}
